Support for multi-platform artifacts when creating builders and buildpack package (#295)
buildpacks/rfcs: text/0000-multiarch-builders-and-package.md

and `--target` flag must be used
- When `platform.os` is not present in [package.toml](https://buildpacks.io/docs/reference/config/package-config/) and `--target` flag is not used, throw a warning messages indicating 
a new `--target` flag is available and how to use it, or some helpful information on how to add `targets` to the `buildpack.toml`
- Keep doing our current process to package a buildpack

### To improve user experience

We propose:
- Add a new `--target` flag with the follwing format `[os][/arch][/variant]:[name@version]` to build for a particular target, once the `platform.os` field is removed, 
this will be the way for end-users to specify the platform for which they want to create single OCI artifact.

- Add `targets` section to `buildpack.toml` to help Buildpack Authors to include support for new platforms without having to update their `pack buildpack package` command in their CI/CD pipelines

- A new folder structure to organize the buildpacks binaries for multi-arch images, similar to this one:
```bash
# Option 1 - no variant is required
.
├── buildpack.toml                 // mandatory

> ├── package.toml                   // mandatory

Reviewer: ```suggestion
├── package.toml                   
```

Reviewer: I don't think package.toml is required today - or at least, it isn't required to live in the buildpack root directory

Looking at this old pack [issue](https://github.com/buildpacks/pack/issues/1082) we might not even need package.toml for component buildpacks if `pack buildpack package` is run inside the buildpack directory containing buildpack.toml.

Taking the above into account, package.toml is probably most useful for composite buildpacks. Following on the discussion of whether composite buildpacks need targets, perhaps we should introduce `[targets]` in package.toml, with the understanding that this is really for composite buildpacks.

Author: I am working on implementing a PoC and I actually got a problem when trying to create a **multi-arch composite buildpack**. We need a place where to define `[targets]` and `buildpack.toml` can't be that place because we don't allow `Orders` and `Targets | Stacks` at the same time.

└── {os}                           // optional
    └── {arch}                     // optional (becomes the platform root folder)
        └── bin
            ├── build              // platform dependant binary (mandatory)
            └── detect             // platform dependant binary (mandatory)

# Option 2 - variant is required
.
├── buildpack.toml                  // mandatory

> ├── package.toml                    // mandatory

Reviewer: ```suggestion
├── package.toml                    
```

and also handle the creation of an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md) will simplify their CI/CD pipelines and make the experience more suitable.

- What is the expected outcome?

`pack buildpack package` and `pack builder create` commands will be updated in a way that they will handle the creation of multi-arch OCI images

# What it is
[what-it-is]: #what-it-is

The end-users for this proposal are **Buildpack authors**, we expect to improve their user experience when creating multi-architecture buildpacks and builders as follows

## Multi-arch example

Let's suppose a **Buildpack author** has a `buildpack.toml` updated to include `targets` as follows:

```toml
# Buildpack API version
api = "0.12"

# Buildpack ID and metadata
[buildpack]
  id = "examples/my-multiarch-buildpack"
  version = "0.0.1"

# List of targets operating systems, architectures and versions

[[targets]]
os = "linux"
arch = "amd64"

[[targets]]
os = "linux"
arch = "arm64"

[[targets]]
os = "windows"
arch = "amd64"

[[targets.distributions]]
name = "windows"

> versions = ["10.0.20348.1970"]

Reviewer 2: I'm wondering what an automated checks around non-code PRs such as a `buildpack.toml` linter (for example) might look like.

> versions = ["10.0.20348.1970"]

Author: > @jjbustamante I believe this should be `[[targets.distros]]` plus `versions` should be `version` (singular) and a string not an array?
> 
> (Plus the same change made to the 3-4 similar references in this file)

Yes!! You are right @edmorley! I will ask how could I do the fix now that this was merged

> versions = ["10.0.20348.1970"]

Author: I just saw you opened a PR to fix a previous RFC, is it possible for you to open one for this?

> versions = ["10.0.20348.1970"]

Reviewer 3: Sure - I've opened #312.

> versions = ["10.0.20348.1970"]

Author: Thank you!! @edmorley

[configuration](https://github.com/opencontainers/image-spec/blob/main/config.md#properties) with the following target 
platform:

```json
{
  "architecture": "",
  "os": "linux"
}
```

After checking the **warning** messages, some end users must feel curious, and the try to use the new `--target` flag.

```bash
pack buildpack package <buildpack> --config ./package.toml --publish --target linux/arm64
Successfully published package <buildpack> and saved to registry

# Or
pack buildpack package  <buildpack> --config ./package.toml --format file --target linux/arm64
Successfully created package <buildpack> and saved to file
```

**Output**: In these cases, pack will create buildpack a package image with the provided binaries and a 
[configuration](https://github.com/opencontainers/image-spec/blob/main/config.md#properties) with the following target 
platform:

```json
{
  "architecture": "arm64",
  "os": "linux"
}
```

> **Important**
> Pack will assume the binaries are appropriate for the given target platform, what the flag is doing is expose a mechanism
> to update the metadata present in the OCI config file

> what about creating a multi-arch image for several target platforms?

Reviewer 4: ```suggestion
what about creating a multi-platform image for several targets?
```

```bash
pack buildpack package <buildpack> --config ./package.toml --publish --target linux/arm64 --target linux/amd64

> A multi-arch buildpack package will be created for target platforms: 'linux/amd64', 'linux/arm64'

Reviewer 4: ```suggestion
A multi-platform buildpack package will be created for: 'linux/amd64', 'linux/arm64'
```

> Successfully published package <buildpack> and saved to registry

Reviewer 4: ```suggestion
Successfully published <buildpack>
```

Author: This is actually the current messaged showed in pack

A multi-arch buildpack package will be created for target platforms: 'linux/amd64', 'linux/arm64'
Successfully created package <buildpack> and saved to file
```

**Output**: two OCI images, with the same binaries, will be created and pushed into the registry, for each image the 
configuration file will be created with the correct `os` and `architecture` and an 
[image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md) will be created to combine them
using the `<buildpack>` name provided. The content of the [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md) 
will be similar to:

```json

{
  "manifests": [
    {
      "digest": "sha256:b492494d8e0113c4ad3fe4528a4b5ff89faa5331f7d52c5c138196f69ce176a6",
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      },
      "size": 424
    },
    {
      "digest": "sha256:2589fe6bcf90466564741ae0d8309d1323f33b6ec8a5d401a62d0b256bcc3c37",
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "platform": {

>         "architecture": "arm",

Reviewer 4: ```suggestion
        "architecture": "arm64",
```

4. Remove `platform.os` support on `pack`

# Drawbacks
[drawbacks]: #drawbacks

- New complexity will be added into `pack`

# Alternatives
[alternatives]: #alternatives

- Do nothing, Buildpack Authors can keep using other tools like`docker buildx imagetools create` or `crane` to update the `architecture` in their Manifest files or 
create [image indexes](https://github.com/opencontainers/image-spec/blob/master/image-index.md).
- Do not deprecate `platform.os` field from [package.toml](https://buildpacks.io/docs/reference/config/package-config/) and add more fields to get the same result, instead of flags
  - I didn't explore this idea

# Prior Art
[prior-art]: #prior-art

- Stack Removal [RFC #096](https://github.com/buildpacks/rfcs/blob/jjbustamante/feature/multi-arch-phase-2/text/0096-remove-stacks-mixins.md)
- This RFC is a continuation of the work started with the proposal to add commands to handle manifest list in pack, see the [RFC](https://github.com/buildpacks/rfcs/pull/283)
- Paketo [RFC #288](https://github.com/paketo-buildpacks/rfcs/pull/288) to publish multi-arch buildpacks

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

> - If I already have an [image indexe](https://github.com/opencontainers/image-spec/blob/master/image-index.md) created and I want to add support for new plaform, how do I do it?

Reviewer 4: ```suggestion
- How would I add support for a new platform to an existing [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md)?
```

- The only values allowed are `linux` and `windows` and by default when is not present, `linux` is being used.
- When exporting to daemon, the `docker.OSType` must be equal to `platform.os`

### To keep compatibility 

We propose:
- Deprecate the `platform.os` field from [package.toml](https://buildpacks.io/docs/reference/config/package-config/). It will be removed after two pack releases with the new feature
- When `platform.os` is present in [package.toml](https://buildpacks.io/docs/reference/config/package-config/), throw a warning messages indicating the field will be removed 
and `--target` flag must be used
- When `platform.os` is not present in [package.toml](https://buildpacks.io/docs/reference/config/package-config/) and `--target` flag is not used, throw a warning messages indicating 
a new `--target` flag is available
- Keep doing our current process to package a buildpack

### To improve user experience

We propose:
- Add a new `--target` flag with the follwing format `[os][/arch][/variant]:[name@version]` to build for a particular target, once the `platform.os` field is removed, 

> this will be the way for end-users to specify the platform for which they want to create single OCI artifact.

Reviewer: I like to think of `--target` as more of a helpful way to narrow down the os/arch when buildpack.toml has multiple definitions, or to avoid creating a manifest list. Is that correct?

Author: Yes, **exactly** when you actually want to avoid creating multiple images, then `target` will help you

            └── detect             // platform dependant binary (mandatory)

# Option 2 - variant is required
.
├── buildpack.toml                  // mandatory
└── {os}                            // optional
    └── {arch}                      // optional
        └── {variant}               // optional
            ├── {name@version-1}    // optional
            │ └── bin
            │     ├── build         // platform dependant binary (mandatory)
            │     └── detect        // platform dependant binary (mandatory)
            └── {name@version-2}    // optional
                └── bin
                    ├── build       // platform dependant binary (mandatory)
                    └── detect      // platform dependant binary (mandatory)
```
> **Note** 
> For cross-compile buildpacks like Paketo, it looks easy to add a step to their Makefile to compile and separate the binaries following this structure. It is important to mention
> that the final buildpack image will not change, this will only change the buildpack structure from `pack` perspective

In case, this folder structure is not suitable for Buildpack Authors, **we propose** a new `path` attribute to be included 

> into the `target` section of the `buildpack.toml` to specify where the binaries are located in the filesystem.

Reviewer: Would that be the location of the `bin` directory, or the buildpack root directory? I think the latter makes more sense, but I could be wrong. Do we need to add this now, or is this something that could be added later if buildpack authors request it?

Author: I was thinking on the **buildpack root directory**, well it came up in a WG meeting suggested by @hone. on the implementation side, I'd propose we could create the ticket and do it later that will simplify the first implementation.

Based on the [RFC-0096](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md) the new `buildpack.toml` schema will look like this:

```toml
[[targets]]
os = "<operating system>"
arch = "<system architecture>"
variant = "<architecture variant>"

# optional
path = "<path to look for the binaries if the folder structure convention is not followed>" 

[[targets.distributions]]
name = "<distribution ID>"
versions = ["<distribution version>"]
```
- When `more than 1 target is defined`
  - When `--publish` or `--format file` is specified
    - For each `target` an OCI image will be created, following these rules
      - `pack` will copy everything in the root folder excluding all `{os}/**` folder 
      - `pack` will determine the **target root binary folder**, this is the specific root folder for a given `target` (based on the `targets.path` in the buildpack.toml or inferring it from a folder structure similar to the one show above)

>         - `pack` will copy everything from the **target root binary folder** into the base buildpack package folder, in case a conflict file name is found, `pack` will **override** the existing file with the latest one

Reviewer 5: Will these two copy steps outlined happen in different image layers? It seems like the idea here is to save disk space for files that are common across architectures. To achieve that, I think you'd need an image layer with the common files, that images for each architecture could include in their layer hierarchy. Maybe this was the idea and I missed it. If so, maybe worth mentioning somewhere.

Otherwise, if common files are in the same image layer as the architecture-specific files, clients would essentially be downloading and storing the common files once for each architecture.

Reviewer 5: Ok, I did find the answer a bit further on: https://github.com/buildpacks/rfcs/pull/295/files#diff-cf9fc6f40b9a670966257d2475741958b49fb15684dd37cf8019514fe779885bR799

Reviewer 5: If the common files are duplicated across architectures anyway, does the common file inheritance have much value? End users of buildpacks and builders won't really get much benefit. And the CNBs I work on (Heroku's), they won't have many (maybe 0) files shared across architectures.

Just thinking common file inheritance is a heavy feature with some potential gotchas, and buildpack authors who want/need to dedupe files could probably get pretty far just by symlinking or hardlinking files into multiple `{os}/{arch}` folders.

Author: Hi @joshwlewis 

Thank you very much for your feedback! I think that at least the `buildpack.toml` and the `package.toml` will be two common files shared across architectures, that's actually a win for a Buildpack Authors, right? because they do not need to keep separate files for each os/arch they want to support.

Reviewer: > common file inheritance is a heavy feature with some potential gotchas

I agree. Maybe it makes sense to take it out of this RFC and add it later if buildpack authors want it

Reviewer: > I think that at least the buildpack.toml and the package.toml will be two common files shared across architectures

I feel like `buildpack.toml` could be the one and only special case. IDK if `package.toml` is something that we even expect to be inside a buildpack root directory at build time. It's just easier to comprehend the packaging step if we don't move too much stuff around inside the directory tree. Perhaps we could spec out that we'd resolve symlinks when creating the layer if folks want to take advantage of that.

Author: Well, if I take a look at some buildpacks for Paketo for example, they have an structure like this one in the root folder:

```bash
├── LICENSE
├── NOTICE
├── README.md
├── buildpack.toml
├── package.toml
├── resources
│ ├── config.properties
```

In these cases, I think they want to shared those files on each platform images right?

Reviewer 5: > In these cases, I think they want to shared those files on each platform images right?

They could copy, symlink, or hard link files into each platform folder. I don't think that's a big challenge for buildpack authors (they probably script this in some way anyway). Honestly, I think it might be a bit more straightforward for them to understand, too. Copy, symlink, and hardlink are established concepts; they won't need to read this RFC or the buildpack spec to know how to do that.

I still think there's convenience in this shared file logic. I'm mostly just thinking you could simplify the initial implementation, long term maintenance, and probably this RFC without it.

```bash
.
└── cnb
    └── buildpacks
        └── {ID}
            └── {version}
                ├── bin
                │    ├── bar        // specific platform binary
                │    ├── build
                │    └── detect
                ├── buildpack.toml
                ├── foo             // override files with the content of the arm64 folder
                └── package.toml
```

#### Buildpacks authors do not use targets AND `platform.os` is not present at `package.toml`

This seems to be the case for [Paketo Buildpacks](https://github.com/paketo-buildpacks/java/blob/main/package.toml) 
or [Heroku](https://github.com/heroku/buildpacks-jvm/blob/main/meta-buildpacks/java/package.toml), and it represents how `pack` 
will work for most the users when new behavior is implemented

A simplified version of Buildpack Authors folder structures is:

```bash
├── bin
│ ├── build
│ └── detect
├── buildpack.toml
└── package.toml
```

In these cases: We expect `pack` to keep doing what is doing today, but with the warning messages we mentioned above to 
let end users know things are changing. 

```bash
pack buildpack package <buildpack> --config ./package.toml --publish 

> Warning: A new '--target' flag is available to set the platform for the buildpack package, using 'linux' as default

Reviewer: ```suggestion
Warning: A new '--target' flag is available to set the platform for the buildpack package, using 'linux/amd64' as default
```

Successfully published package <buildpack> and saved to registry

# Or
pack buildpack package  <buildpack> --config ./package.toml --format file 

> Warning: A new '--target' flag is available to set the platform for the buildpack package, using 'linux' as default

Reviewer: ```suggestion
Warning: A new '--target' flag is available to set the platform for the buildpack package, using 'linux/amd64' as default
```

Successfully created package <buildpack> and saved to file
 ```
**Output**: pack will create a buildpack package image (as it is doing it today) with the provided binaries and a 
[configuration](https://github.com/opencontainers/image-spec/blob/main/config.md#properties) with the following target 
platform:

```json
{

>   "architecture": "",

Reviewer: This is how it works today, but this feels like a bug. The OCI image spec says this field is required.

Author: Maybe it is a bug, but apparently nobody is complaining about it! I didn't want to start filling this attribute and break the current behavior for end-users.

Reviewer: Actually this did come up recently ([here](https://github.com/buildpacks/pack/issues/1968) and [here](https://cloud-native.slack.com/archives/C0331B61A1Y/p1706546411819879)) I __think__ we said this RFC would fix those cases but maybe we assumed first the buildpack author would upgrade to use `targets`. In any case, it probably doesn't hurt... I'm fine to leave it out of this RFC though

Author: Yeah, I think is what @jericop is trying to fix also [here](https://github.com/buildpacks/pack/pull/2080)

### Examples

Let's use some examples to explain the expected behavior in different use cases

#### `Targets` are not present in `builder.toml`

This is probably the case for most of the Buildpack Authors, for example 
[Paketo](https://github.com/paketo-buildpacks/builder-jammy-tiny/blob/main/builder.toml), lets suppose a`buildpack.toml` 
like:

```toml
# Buildpacks to include in builder
[[buildpacks]]
uri = "<some buildpacks>"

# Order used for detection
[[order]]
[[order.group]]
id = "<some order>"
version = "<some version>"

[stack]
id = "io.buildpacks.samples.stacks.jammy"
build-image = "cnbs/sample-base-build:jammy"
run-image = "cnbs/sample-base-run:jammy"
```
Or we include `build` and `run` images

```toml
# Base images used to create the builder
[build]
image = "cnbs/sample-base-build:jammy"
[run]
[[run.images]]
image = "cnbs/sample-base-run:jammy"
```

In these cases, the expected output will be similar to:

```bash
pack builder create <builder> --config ./builder.toml 
Warning: "stack" has been deprecated, prefer "targets" instead: https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md

> Warning: A new '--target' flag is available to set the target platform for the builder, using 'linux/amd64' as default

Reviewer: This assumes that the build image is a manifest list, no? Because if the build image is linux/arm64 (only) then we create an arm64 image, no?

Author: Yeah, the `pack builder create` command assumes the build, run and buildpack images are multi-arch, and when building for a particular platform, `pack` should validate it

[meta]: #meta
- Name: Multi-arch support for builders and buildpack packages
- Start Date: 2023-09-14
- Author(s): @jjbustamante
- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)

# Summary
[summary]: #summary

The problem for adding support for multi-arch buildpacks can be divided into three parts:
1. Support buildpack authors to **migrate their existing buildpacks** to support multi-arch.
2. Support buildpack authors to **create new buildpacks and builders** that handle multi-arch from the beginning.
3. Support application developers to **create application images** using multi-arch buildpacks and builders.

The purpose of this RFC is to solve the statement 2, adding the capability to the commands:

- `pack buildpack package`
- `pack builder create`

to create individuals OCI images artifacts per each os and arch (builders and buildpack packages) and handle the creation for the  [image index,](https://github.com/opencontainers/image-spec/blob/master/image-index.md)

> that combines them into one single consumable tag for end-users.

Reviewer 6: To confirm, this will support `--publish` of the generated OCI images also? We'd been creating OCI archives in some of our pipelines and when we started generating multi-arch images in OCI archive files, it broke the tools we were using to move things around (`crane` and `skopeo`). It'd be really helpful to have that support in the tool.

Author: Yes, the idea is to push the generated OCI images and create the Image Index

For those buildpack authors that are using `cross-compile` languages like [go](https://go.dev/) or [rust](https://www.rust-lang.org/) or maybe bash scripts, adding the capability to `pack buildpack package` and `pack builder create` to create multi-arch images 
and also handle the creation of an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md) will simplify their CI/CD pipelines and make the experience more suitable.

- What is the expected outcome?

`pack buildpack package` and `pack builder create` commands will be updated in a way that they will handle the creation of multi-arch OCI images

# What it is
[what-it-is]: #what-it-is

The end-users for this proposal are **Buildpack authors**, we expect to improve their user experience when creating multi-architecture buildpacks and builders as follows

## Multi-arch example

Let's suppose a **Buildpack author** has a `buildpack.toml` created with a content similar to this one:

```toml
# Buildpack API version
api = "0.12"

# Buildpack ID and metadata
[buildpack]
  id = "examples/my-multiarch-buildpack"
  version = "0.0.1"

# List of targets operating systems, architectures and versions

> [[targets]]

Reviewer: I am sorry for opening this can of worms, but technically "target" when used in the context of say `CNB_TARGET_OS` or `CNB_TARGET_ARCH` refers exclusively to the run image. Right now, all target data for the build image and the run image must match - but perhaps that won't always be the case? What will we do then?

Reviewer 7: @natalieparellano what do you think we should use here?

Reviewer: I am pretty sure that the current architecture will accommodate this. What would probably end up happening is that buildpack.toml would have a list of `[[targets]]` and the buildpack os/arch/bin directories would be a subset of that list. Then at build time (somehow, but probably in run.toml) you'd have a list of run images to fulfill all the os/arch combinations in `[[targets]]` and by iterating through that list, you have a multi-arch application image. The key is that the mapping between buildpack.toml `[[targets]]` and the buildpack os/arch/bin directories would not necessarily be 1:1, whereas today we assume it is.

os = "linux"
arch = "amd64"

[[targets]]
os = "linux"
arch = "arm64"

[[targets]]
os = "windows"
arch = "amd64"

[[targets.distributions]]
name = "windows"
versions = ["10.0.20348.1970"]
```

And organizes the binaries according to their os/arch with a structure similar to this one:

```bash
my-multiarch-buildpack
├── buildpack.toml

> ├── linux-amd64

Reviewer: Is there any case for doing something like

```
└── linux
    └── amd64
        └── bin
            ├── build
            └── detect
```

or (in the case that the buildpack runs on any arch)

```
└── linux
    └── bin
        ├── build
        └── detect
```

with both being accepted? Asking because that pairs nicely with the command line input `[os][/arch][/variant]:[name@version]`. Having to type linux-amd64 sometimes and linux/amd64 other times might be confusing.

Author: I agreed, I created the structure first and then I thought on the `flag`. I will change it

│         └── bin
│             ├── build
│             └── detect
├── linux-arm64
│         └── bin
│             ├── build
│             └── detect
└── windows-amd64
    └── 10.0.20348.1970
        └── bin
            ├── build.bat
            └── detect.bat
```

Now `pack` will be able to package them separately for each os/arch family, following our [guide](https://buildpacks.io/docs/buildpack-author-guide/package-a-buildpack/) 
we will need to create a `package.toml` file.

The `package.toml` file will be similar to the one describe in our documentation:

```toml
[buildpack]
uri = "examples/my-multiarch-buildpack"
# OR a .tgz with the previous folder structure
uri = "my-multiarch-buildpack.tgz"
```

But in this case, we actually **remove** the [platform](https://buildpacks.io/docs/reference/config/package-config/) section because it will be taken from the `buildpack.toml`.

Packaging a multi-arch buildpack will require the output to be **publish** to a registry or **saved on disk** in OCI layout format.

```bash
pack buildpack package my-buildpack --config ./package.toml --publish --multi-arch
# Or

> pack buildpack package my-buildpack.cnb --config ./package.toml --format file --multi-arch

Reviewer 6: Could the `--multi-arch` flag be inferred? For example, if `buildpack.toml` contains multiple target/arch's then this flag is automatic. Or are we worried someone might have that but still only want to build a buildpack for a single arch?

Author: It is exactly this: 
> Or are we worried someone might have that but still only want to build a buildpack for a single arch?
At first, I was thinking on doing it without the `--multi-arch` flag, but then I decided to force the user to give the acknowledgment

Reviewer: I agree that having to type `--multi-arch` everywhere is cumbersome. I think if the buildpack author took the trouble to arrange their binaries according to the nested directory structure then we should just infer that they want a multi-arch image. Or maybe there could be a `pack config` option that sets multi-arch (or not) as the default.

Author: I removed the `--multi-arch` flag

```

In these cases each `target` entry corresponds to a different buildpack image that is exported into an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md)

### Package a multi-arch Builder

In case of packing a **Builder**, we assume the following premises: 

> 1. Buildpack authors migrated their `builder.toml` and [remove the stack](https://github.com/buildpacks/pack/issues/1303) concept.

Reviewer 6: I'm concerned about this. If we make the change in #1303, will we still be able to support stacks? We have a lot of users in production. We need to support both for a while so we can give folks plenty of time to migrate.

If this requires removing and stopping support for stacks then I suspect that this will hold us up for a while on being able to use this functionality.

Is there any way we could change this to something like `Buildpack authors migrated `builder.toml` and have added the entries to `builder.toml` for the new run/build image format.`. So that we could have both for some period of time as we migrate?

Author: Let me double check this, I need the targets specify in the `builder.toml`. If the users still specify stacks for compatibility I think we can workaround it.

Reviewer: I think for the purposes of this RFC any `[[stacks]]` in builder.toml can just be ignored.

Author: Yeah, I think stacks can be ignored, and we will follow the steps described in the [plan](https://docs.google.com/document/d/1bExpr31U5R5yQ6fncpl5YcdosWVwYcXgkt12vE-lpvU/edit#heading=h.jpmej4uevvh8) to keep compatibility

image = "index.docker.io/paketobuildpacks/run-jammy-tiny:latest"

[build]
# This image is used at build-time, in case of multi-arch images it must point to Image Index
image = "docker.io/paketobuildpacks/build-jammy-tiny:0.2.3"

[[targets]]
os = "linux"
arch = "amd64"

[[targets]]
os = "linux"
arch = "arm64"
```

As we can see, the proposal is based on the assumption that the `run-image`, `build-image` and `buildpacks` to include 
in the builder are **multi-arch artifacts**, and we can reach them by reading an 
[image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md)

```bash
pack builder create my-jammy-builder --config ./builder.toml \ 
      --multi-arch \ 
      --publish
```

In this case, because `targets` are specified and `--multi-arch` is being used `pack` will follow the builder creation process for **each provided platform**, 
pulling the correct (based on os/arch) buildpacks, build and run images and creating different builders images that are 
exported and combined into an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md)

# How it Works
[how-it-works]: #how-it-works

## Buildpack Package

> As a quick summary, our current process to create a buildpack package involves:

Reviewer 6: Would anything change for a composite/meta buildpack? I'd assume that you would need to have multi-arch images for every buildpack in the composite buildpack. Otherwise, would it work the same as a regular component buildpack?

Author: That's correct, I will add a new example with the composite builpack, but we are assuming you'd need to have multi-arch images for every buildpack in the composite buildpack

- The end-users defined `os` for the OCI image using the [package.toml](https://buildpacks.io/docs/reference/config/package-config/) when they want to export a buildpack package.
- The only values allowed are `linux` and `windows` and by default when is not present, `linux` is being used.
- When exporting to daemon, the `docker.OSType` must be equal to `platform.os`

### To keep compatibility 

We propose:
- Deprecate the `platform.os` field from [package.toml](https://buildpacks.io/docs/reference/config/package-config/). It will be removed after two pack releases with the new feature
- When `platform.os` is present in [package.toml](https://buildpacks.io/docs/reference/config/package-config/), throw a warning messages indicating the field will be removed 

> and `--platform` flag must be used

Reviewer: This was brought up elsewhere, but `--platform` might be confusing when the corresponding word in buildpack.toml and builder.toml is `--target`.

- When `platform.os` is not present [package.toml](https://buildpacks.io/docs/reference/config/package-config/) and `--platform` flag is not used, throw a warning messages indicating 
a new `--platform` flag is available
- Keep doing our current process to package a buildpack

### To improve user experience

We propose:
- Add a new `--platform` flag with format `[os][/arch][/variant]:[name@version]` to build for a particular target, once the `platform.os` field is removed, 
this will be the way for end-users to specify the platform for which they want to create single OCI artifact.
> **Note**
> We want to avoid duplicated information, we expect to have `targets` with all the platform supported by a buildpack

- Add a new boolean `--multi-arch` flag to indicate pack it must create multiples OCI artifacts and combine them with an [image index](https://github.com/opencontainers/image-spec/blob/master/image-index.md), 
this flag must be used in conjunction with `--publish` or `--format file` flags and error out when `daemon` is selected.
> **Note**

> > I got feedback to avoid creating an Image Index if users are not expecting that, the flag is an acknowledgment for creating the index 

Reviewer 6: Similar thoughts here. Could this be inferred based on the presence of multiple targets? I get that we don't want to just suddenly start making image indexes as that changes things. But couldn't we only do this if multiple targets with different arch's are specified? Or is there a case where you'd have multiple target/arch's and still not want that behavior?

Reviewer: I am trying to think of who would be specifying multiple targets today and I'm guessing it would just be bash buildpacks. Maybe we could print a warning if multiple targets are defined but there is only one ./bin/detect or ./bin/build.

Author: The scenarios I was thinking were these:
1. I do not have `targets` on my `buildpack.toml` 
   - In this case, the idea is to `pack buildpack package` do the same thing as it is doing today. 
1. I do have `targets` on my `buildpack.toml` 
  - **Daemon** case: `pack` should build my **buildpack package** with the platform defined by the daemon (as it is doing today) without requiring any new flag. 
  - **Registry or File** case: 
     -  If more than 1 target is defined: `pack` will create the `Image Index`, If I just want to create 1 single package, I will need to use `--target` flag to defined which one.
     - If only 1 target is available, nothing changes

- A new folder structure to organize the buildpacks binaries for multi-arch images
```bash
.
├── buildpack.toml
└── {os}-{arch}             // optional
    └── {variant}           // optional
        ├── {version-1}     // optional
        │   └── bin
        │       ├── build
        │       └── detect
        └── {version-2}     // optional
            └── bin
                ├── build
                └── detect
```
> **Note** 

> > For cross-compile buildpacks like Paketo, it looks easy to add a step to their Makefile to compile and separate the binaries following this structure.

Reviewer 6: Yes, that shouldn't be an issue.

Would this only change the buildpacks structure presented to `pack` or would it also change the buildpack structure in the actual image that's generated? The latter would require buildpack API/spec changes, right?

Author: This will only change the buildpack structure presented to `pack`, the final image should keep the structure as we defined in the spec.

> 

Trying to use the new flags

```bash
pack buildpack package <buildpack> --config ./package.toml --publish --multi-arch
Error: 'targets' or 'platforms' must be defined when creating a multi-architecture buildpack

# Or
pack buildpack package  <buildpack> --config ./package.toml --format file --multi-arch
Error: 'targets' or 'platforms' must be defined when creating a multi-architecture buildpack
```

In these cases, we don't have enough information to create a multi-arch buildpack, and pack must fail its execution. 

A valid way to re-write the command could be:

```bash
pack buildpack package <buildpack> --config ./package.toml --publish --platform linux/arm64 --platform linux/amd64 --multi-arch
A multi-arch buildpack package will be created for platforms: 'linux/amd64', 'linux/arm64'
Successfully published package <buildpack> and saved to registry

# Or
pack buildpack package  <buildpack> --config ./package.toml --format file --platform linux/arm64 --platform linux/amd64  --multi-arch
A multi-arch buildpack package will be created for platforms: 'linux/amd64', 'linux/arm64'
Successfully created package <buildpack> and saved to file

> ```

Reviewer 6: So this command could be used if one hasn't updated to use targets yet? i.e. still using stacks?

What about buildpacks that are migrating from stacks to no-stacks and have both present? This use case concerns me the most, cause I know we (Paketo) will be in this state for a while.

Reviewer 6: I'd be interested to hear more in general about this state of migrating from stacks to no-stacks. What commands would use the old stack metadata? what commands would use the new target info? I'm assuming we'd keep the two synchronized, but just curious more about how this would all work.

Or am I overthinking this, and the net result (i.e. the images generated) will be the same regardless of stacks or no stacks?

Author: Good point, regarding the stack removal [plan](https://docs.google.com/document/d/1bExpr31U5R5yQ6fncpl5YcdosWVwYcXgkt12vE-lpvU/edit#heading=h.jpmej4uevvh8) we are planning to support both scenarios (stacks and run, build images), based on your feedback the idea will be to clarify how a world with:
- stacks
- targets
- multi-arch support 
will work
